drivers/segger/console_rtt.c

```diff
@@ -0,0 +1,552 @@
+ * Name: rtt_serial_ioctl
+ ****************************************************************************/
+
+static int rtt_serial_ioctl(FAR struct file *filep, int cmd,
+                            unsigned long arg)
+{
+  int ret;
+  FAR struct inode *inode;
+  FAR struct rtt_serial_dev_s *dev;
+  DEBUGASSERT(filep != NULL && filep->f_inode != NULL);
+  inode = filep->f_inode;
+  dev   = inode->i_private;
+
+  switch (cmd)
+    {
+      /* Get the number of bytes that are immediately available for
+       * reading from the source pipe.
+       */
+
+    case FIONREAD:
+      {
+        ret = SEGGER_RTT_HasData(dev->channel);
+      }
+      break;
+
+      /* Get the number of bytes waiting in the sink pipe (FIONWRITE)
+       * or the number of unused bytes in the sink pipe (FIONSPACE).
+       */
+
+    case FIONWRITE:
+      {
+        ret = SEGGER_RTT_GetBytesInBuffer(dev->channel);
+      }
+
+      /* Get the number of free bytes in the TX buffer */
+
+    case FIONSPACE:
+      {
+        ret = SEGGER_RTT_GetAvailWriteSpace(dev->channel);
+      }
+      break;
+    default:
+      {
+        ret = -ENOTTY;
+      }
+      break;
+    }
+
+  return ret;
+}
+
+/****************************************************************************
+ * Name: rtt_serial_write
+ ****************************************************************************/
+
+static ssize_t rtt_serial_write(FAR struct file *filep,
+                                FAR const char *buffer, size_t buflen)
+{
+  FAR struct inode *inode;
+  struct rtt_serial_dev_s *dev;
+  DEBUGASSERT(filep != NULL && filep->f_inode != NULL);
+  inode = filep->f_inode;
+  dev   = inode->i_private;
+  return SEGGER_RTT_Write(dev->channel, buffer, buflen);
+}
+
+void rtt_serial_register(FAR struct rtt_serial_dev_s *dev)
+{
+  if (dev->channel != 0)
+    {
+      FAR char *rxbuf = (FAR char *)kmm_malloc(dev->rxsize);
+      FAR char *txbuf = (FAR char *)kmm_malloc(dev->txsize);
+      SEGGER_RTT_ConfigUpBuffer(dev->channel, dev->devname, txbuf,
+                                dev->txsize,
+                                SEGGER_RTT_MODE_NO_BLOCK_SKIP);
+      SEGGER_RTT_ConfigDownBuffer(dev->channel, dev->devname, rxbuf,
+                                  dev->rxsize,
+                                  SEGGER_RTT_MODE_NO_BLOCK_SKIP);
+    }
+
+  sinfo("Registering channel %d, devname:%s\n", dev->channel, dev->devname);
+  register_driver(dev->devname, &g_rtt_serial_ops, 0666, (void *)dev);
+}
+
+#endif /* CONFIG_RTT_SERIAL */
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: up_uart
+ ****************************************************************************/
+
+void rtt_serial_initialize(void)
+{
+  int i;
+  UNUSED(i);
+
+#ifdef CONFIG_SEGGER_RTT
```

Comment:
why need check CONFIG_SEGGER_RTT which should be enable before we can reach here.